Add challenge and project creation and editing forms

fluix-dev opened the pull request:
TODO:
 - [x] Button for creation
 - [x] Button for editing
 - [x] Permission checks for editing
 - [ ] ~~Only display certain types of tags for addition~~ See #36 
 - [x] Fix Chosen tag dropdown

And in a separate PR:
 - [ ] Convert profile editing to `base_form.html`
 - [ ] Normalize URL naming conventions
 - [ ] Alphabetize Python imports
 - [ ] Make description mandatory


Fixes #16 
Fixes #17 
Fixes #19

fluix-dev committed:
Add challenge and project creation and editing forms

fluix-dev committed:
Switch to Jinja2

This also cleans up some template styling.

fluix-dev committed:
Add buttons and permission checks

fluix-dev commented:
@dulldesk or @skyflaren could one of you take a look at the CSS issue.

dulldesk commented on app/templates/base_form.html:
```diff
@@ -0,0 +1,59 @@
+    </div>
+    <form action="." method="POST">
+      {% csrf_token %}
+      <div class="frame__body">
```
```suggestion
      <div class="frame__body" style="overflow-y:initial">
```
Prevents the tags selection dropdown from overflowing the div and making it scroll.

fluix-dev replied:
What happens instead?

**Edit**: Just tested, looks really nice :+1:

dulldesk replied:
The drop down list overlays on top of whatever happens to be below it (in this case, the footer). No extra space is created.

dulldesk commented:
Is it possible to force the challenge start/end fields to render as time inputs?

dulldesk commented:
Forms need to redirect to a URL upon completion. 
> ImproperlyConfigured at /challenge/create/
No URL to redirect to.  Either provide a url or define a get_absolute_url method on the Model.

Perhaps they could redirect to their creation?

dulldesk committed:
Do not add left padding to iconless input field

Only muliple selection fields that have an icon and are affected by Chosen (which does not add space by default) will have space added for the icon.

dulldesk commented:
9f90968 removes the padding that was added for the icon in `explore` but is unnecessary in inputs that don't add an icon in the input bar (i.e. in the challenge/project creation/edit forms).

dulldesk commented:
Perhaps add a note somewhere in the form that required fields are marked blue.

fluix-dev commented:
> Perhaps add a note somewhere in the form that required fields are marked blue.

I was thinking of adding `(Optional)` to the time fields. Some others are _technically_ optional but we want them filled out.

**Edit**: I don't want to do this anymore. `help_text`s shouldn't have this kind of information. Changing color also aligns with what Django's admin does.

dulldesk commented:
That could work. Would the label colouring still be needed?

fluix-dev commented:
> That could work. Would the label colouring still be needed?

Nah, not really.

**Edit**: I'm a liar, see above.

fluix-dev committed:
Fix tag selection overflow

fluix-dev committed:
Add success urls

fluix-dev committed:
Abstract update and creation view classes

fluix-dev committed:
Add placeholders to datetime fields

fluix-dev committed:
Make it clear that blue fields are required

fluix-dev committed:
Add project creation for challenge endpoint

fluix-dev committed:
Add back removed indentation

dulldesk commented:
We should probably add a warning somewhere that if you remove your account as a `creator`, then you lose `creator` permissions. (Or, alternately, add a modal warning if a user is trying to do such a thing). 

Should we also automatically add the user as a `creator` for a project/challenge they are trying to create, e.g. by adding the user account to the `creators` input field (and they can choose to remove it)?

dulldesk commented:
Will permissions for creating challenges be restricted?

dulldesk commented:
Upon creating a project for a challenge this error is thrown: 
> NameError at /project/create/<challenge pk>/
> name 'get_success_url' is not defined

fluix-dev commented:
> We should probably add a warning somewhere that if you remove your account as a creator, then you lose creator permissions. (Or, alternately, add a modal warning if a user is trying to do such a thing).

Yeah I'll add this into the `help_text`. You actually can't remove all creators, so at least one person will always be able to edit.

> Should we also automatically add the user as a creator for a project/challenge they are trying to create, e.g. by adding the user account to the creators input field (and they can choose to remove it)?

I think the easiest way to do this would be through a bit of JS, so yeah I'll do it.

> Will permissions for creating challenges be restricted?

I'm not sure, but it's probably a good idea. I initially thought everyone can until they are "muted" or do something silly, but limiting challenge creation to only specific people in the first place is probably a good idea.

fluix-dev commented:
@Gongsta what do you think about challenge creation permissions? Should everyone be able to create a challenge or only people that contact us and get the permission?

fluix-dev committed:
Merge branch 'master' of github.com:KuriusMTL/stratus into master

fluix-dev committed:
Required permission to create challenges

fluix-dev committed:
Clarify what removing yourself as creator does

fluix-dev committed:
Blackify urls

fluix-dev commented:
Alright, I think I fixed everything.

fluix-dev committed:
Add logged in user as a creator by default

dulldesk commented:
Should we add a notice if a challenge has been closed, e.g. in the place of the `Create Project` (for a challenge) button? The user may not realise the timestamps' presence, especially as they are given equal emphasis to (e.g.) tags.

dulldesk reviewed (changes requested):
This issue does not seem to have been resolved yet. https://github.com/KuriusMTL/HackItForward/pull/33#issuecomment-694245133

fluix-dev commented:
> This issue does not seem to have been resolved yet. #33 (comment)

Fixed, and improved in general.

fluix-dev committed:
Improve success url resolution

fluix-dev committed:
Update migration dependency

fluix-dev committed:
Merge branch 'master' of github.com:KuriusMTL/stratus into master

fluix-dev committed:
Update migrations

dulldesk commented on app/models.py:
```diff
@@ -170,6 +173,20 @@ def short_creators(self):
+    def get_edit_url(self):
+        return reverse("challenge_edit", args=[self.pk])
+
+    def is_open(self):
+        return self.start <= timezone.now() <= self.end
+
```
An error is thrown when the challenge timestamps are blank

fluix-dev committed:
Add challenge start and end time validation

dulldesk commented on app/models.py:
```diff
@@ -167,6 +168,15 @@ class Challenge(models.Model):
+        if self.start and not self.end:
+            raise ValidationError("An end date must be provided if you have a start date.")
+        if not self.start and self.end:
+            raise ValidationError("An start date must be provided if you have an end date.")
```
```suggestion
            raise ValidationError("A start date must be provided if you have an end date.")
```
Typo

fluix-dev committed:
Fix typo

Co-authored-by: dulldesk <54601393+dulldesk@users.noreply.github.com>

fluix-dev committed:
Remove useless super() call

dulldesk commented:
[diff](https://github.com/KuriusMTL/HackItForward/pull/33/files#diff-b5099d1f0106025f31781ccbe0a59d13R163-R166)

`context["time_labels"]` may display null timestamps, should anything be done about this?

fluix-dev commented:
This should have been covered by https://github.com/KuriusMTL/HackItForward/issues/11

dulldesk commented:
Timestamps were not blankable then but that is valid.

fluix-dev commented:
Fair, then it should have been https://github.com/KuriusMTL/HackItForward/pull/40

fluix-dev commented:
It's super simple, but not related to this, so let's just commit it after.
